zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/client/FederatedZkClient.java

```diff
@@ -20,344 +20,486 @@
-    return null;
+    return getZkClient(path).deserialize(data, path);
   }
 
   @Override
   public void setZkSerializer(ZkSerializer zkSerializer) {
-
+    _pathBasedZkSerializer = new BasicZkSerializer(zkSerializer);
+    _zkRealmToZkClientMap.values()
+        .forEach(zkClient -> zkClient.setZkSerializer(_pathBasedZkSerializer));
   }
 
   @Override
   public void setZkSerializer(PathBasedZkSerializer zkSerializer) {
-
+    _pathBasedZkSerializer = zkSerializer;
+    _zkRealmToZkClientMap.values().forEach(zkClient -> zkClient.setZkSerializer(zkSerializer));
   }
 
   @Override
   public PathBasedZkSerializer getZkSerializer() {
-    return null;
+    return _pathBasedZkSerializer;
+  }
+
+  private String create(final String path, final Object dataObject, final List<ACL> acl,
+      final CreateMode mode, final String expectedSessionId) {
+    if (mode.isEphemeral()) {
+      throwUnsupportedOperationException();
+    }
+
+    // Create mode is not session-aware, so the node does not have to be created
+    // by the expectedSessionId.
+    return getZkClient(path).create(path, dataObject, acl, mode);
+  }
+
+  private ZkClient getZkClient(String path) {
+    // If FederatedZkClient is closed, should not return ZkClient.
+    if (isClosed()) {
+      throw new IllegalStateException(FEDERATED_ZK_CLIENT + " is closed!");
+    }
+
+    String zkRealm = getZkRealm(path);
+    if (!_zkRealmToZkClientMap.containsKey(zkRealm)) {
+      // Synchronized to avoid creating duplicate ZkClient for the same ZkRealm.
+      synchronized (_zkRealmToZkClientMap) {
+        if (!_zkRealmToZkClientMap.containsKey(zkRealm)) {
+          _zkRealmToZkClientMap.put(zkRealm, createZkClient(zkRealm));
+        }
+      }
+    }
+
+    return _zkRealmToZkClientMap.get(zkRealm);
+  }
+
+  private String getZkRealm(String path) {
+    String zkRealm;
+    try {
+      zkRealm = _metadataStoreRoutingData.getMetadataStoreRealm(path);
+    } catch (NoSuchElementException ex) {
+      throw new NoSuchElementException("Cannot find ZK realm for the path: " + path);
+    }
+
+    if (zkRealm == null || zkRealm.isEmpty()) {
+      throw new NoSuchElementException("Cannot find ZK realm for the path: " + path);
+    }
+
+    return zkRealm;
+  }
+
+  private ZkClient createZkClient(String zkAddress) {
+    LOG.debug("Creating ZkClient for realm: {}.", zkAddress);
+    return new ZkClient(new ZkConnection(zkAddress), (int) _clientConfig.getConnectInitTimeout(),
+        _clientConfig.getOperationRetryTimeout(), _pathBasedZkSerializer,
+        _clientConfig.getMonitorType(), _clientConfig.getMonitorKey(),
+        _clientConfig.getMonitorInstanceName(), _clientConfig.isMonitorRootPathOnly());
+  }
+
+  private void throwUnsupportedOperationException() {
+    throw new UnsupportedOperationException(
+        "Session-aware operation is not supported by " + FEDERATED_ZK_CLIENT
+            + ". Instead, please use " + DEDICATED_ZK_CLIENT_FACTORY + " for this operation");
```

Comment:
Nit: Should we tell people to use DedicatdZkClient instead of DedicatedZkClientFactory?

Reply:
`DedicatedZkClientFactory` is what we provide and recommend users to create a RealmAwareZkClient which is a `DedicatedZkClient` instance. `DedicatedZkClient ` is protected and we don't want users to directly use it, right?

Reply:
I see what you're saying, but the way it's phrased here may confuse users.

Say this:

"Instead, please use DedicatedZkClientFactory to create an instance of dedicated RealmAwareZkClient for ephemeral operations".

This is because we don't use the factory for ephemeral operations.